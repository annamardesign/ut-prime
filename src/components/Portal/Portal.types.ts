import {WithStyles, withStyles, createStyles} from '@material-ui/core/styles';
import React from 'react';

import logo from '../images/logo.png';

export interface Props {
}

const styles = createStyles({
    '@global': {
        html: {
            fontSize: 14
        }
    },
    headerContainer: {
        zIndex: 2,
        cursor: 'default'
    },
    headerTitle: {
        paddingLeft: 16,
        paddingRight: 16,
        fontSize: 15,
        fontWeight: 700,
        lineHeight: '16px'
    },
    headerLogo: {
        margin: '0 0px 0 10px',
        width: 36,
        height: 36,
        background: `url(${logo})`,
        backgroundSize: 'contain',
        cursor: 'pointer'
    },
    tabs: {
        display: 'flex',
        flexDirection: 'column',
        flexGrow: 1,
        '& .p-tabview-panels': {
            padding: 0,
            flexGrow: 1,
            display: 'flex',
            flexDirection: 'column'
        },
        '& .p-tabview-panel': {
            flexGrow: 1
        }
    },
    menuGrow: {
        border: 0,
        flexGrow: 1
    },
    menu: {
        border: 0
    }
});

export const Styled = withStyles(styles);
export type StyledType = React.FC<Props & WithStyles<typeof styles>>
