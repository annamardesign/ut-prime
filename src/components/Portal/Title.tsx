import React from 'react';
import { useDispatch } from 'react-redux';

const Title = ({data}) => {
    const dispatch = useDispatch();
    return (
        <>
            <span>
                {data.title}
            </span>
            <i
                className='dx-icon dx-icon-close'
                style={{
                    display: 'inline-block',
                    opacity: 0.6,
                    marginRight: 0,
                    marginLeft: 7,
                    fontSize: 18
                }}
                onClick={() => dispatch({
                    type: 'front.tab.close',
                    data
                })}
            />
        </>
    );
};

export default Title;
