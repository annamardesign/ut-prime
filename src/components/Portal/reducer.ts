const handlers = {
    'front.tab.show'(state, {title, path, component}) {
        return {
            ...state,
            tabs: [...(state.tabs || []), {
                title: title,
                path: path,
                component
            }]
        };
    },
    'front.tab.close'(state, {index}) {
        return {
            ...state,
            tabs: (items => {
                items.splice(index, 1);
                return items;
            })([...(state.tabs || [])])
        };
    }
};

export default (state = {}, action) => {
    const handler = handlers[action.type];
    return handler ? handler(state, action) : state;
};
