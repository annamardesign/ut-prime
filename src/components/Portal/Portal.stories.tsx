import React from 'react';
import { withReadme } from 'storybook-readme';
import type { Story, Meta } from '@storybook/react';

// @ts-ignore: md file and not a module
import README from './README.md';
import Portal from './index';
import type { Props } from './Portal.types';
import Explorer from '../Explorer';

const meta: Meta = {
    title: 'Portal',
    component: Portal,
    decorators: [withReadme(README)],
    args: {
        state: {}
    }
};
export default meta;

const Template: Story<Props & {state: {}}> = ({state, ...args}) => <Portal {...args} />;

export const Basic: Story<Props> = Template.bind({});
Basic.args = {
    state: {
        portal: {
            tabs: [{
                title: 'Tab 1',
                path: '/tab1',
                Component() {
                    return (
                        <Explorer
                            fetch={() => Promise.resolve({
                                items: [...Array(50).keys()].map(number => ({
                                    id: number,
                                    name: `Item ${number}`,
                                    size: number * 10
                                }))
                            })}
                            keyField='id'
                            resultSet='items'
                            schema={{
                                properties: {
                                    name: {
                                        title: 'Name'
                                    },
                                    size: {
                                        title: 'Size'
                                    }
                                }
                            }}
                            columns = {['name', 'size']}
                            details={{
                                name: 'Name'
                            }}
                        >
                            <div>Navigation component</div>
                        </Explorer>
                    );
                }
            }, {
                title: 'Tab 2',
                path: '/tab2',
                Component() { return <div>tab 2 body</div>; }
            }]
        }
    }
};

export const ErrorTab: Story<Props> = Template.bind({});
ErrorTab.args = {
    state: {
        portal: {
            tabs: [{
                title: 'Error',
                path: '/error',
                Component() {
                    throw new Error('Error message');
                }
            }]
        }
    }
};
