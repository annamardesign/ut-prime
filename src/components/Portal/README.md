# Portal

`Portal` shows application pages in a tabbed UI.
It must be wrapped in a Redux Provider, where
the information for the active tabs is stored.

## How to use

```jsx
import {Provider} from 'react-redux';
import Portal from 'ut-front-devextreme/core/Portal';

<Provider store={store}>
    <Portal />
</Provider>
```

See the [storybook source](./Portal.stories.tsx) for examples.

## Props

| propName  | propType | defaultValue | isRequired |
| --------- | -------- | ------------ | ---------- |
