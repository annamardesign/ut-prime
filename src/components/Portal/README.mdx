# Portal
import { Canvas, Story, ArgsTable } from "@storybook/addon-docs";

`Portal` shows application pages in a tabbed UI.
It must be wrapped in a Redux Provider, where
the information for the active tabs is stored.

## How to use

```jsx
import {Provider} from 'react-redux';
import Portal from 'ut-prime/core/Portal';

<Provider store={store}>
    <Portal />
</Provider>
```

<Canvas>
    <Story id="portal--basic" />
</Canvas>

## Properties

<ArgsTable story="."/>
