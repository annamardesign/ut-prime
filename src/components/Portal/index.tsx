import React from 'react';
import Box, {Item} from 'devextreme-react/box';
import { useSelector, useDispatch } from 'react-redux';
import TabPanel from 'devextreme-react/tab-panel';
import { useLocation } from 'react-router-dom';

import Menu from 'devextreme-react/menu';

import Context from '../Context';
import Async from '../Async';
import Text from '../Text';

import Title from './Title';
import { Styled, StyledType } from './Portal.types';

// https://usehooks.com/useWindowSize/
function useWindowSize() {
    // Initialize state with undefined width/height so server and client renders match
    // Learn more here: https://joshwcomeau.com/react/the-perils-of-rehydration/
    const [windowSize, setWindowSize] = React.useState({
        width: undefined,
        height: undefined
    });

    React.useEffect(() => {
        function handleResize() {
            setWindowSize({
                width: window.innerWidth,
                height: window.innerHeight
            });
        }

        window.addEventListener('resize', handleResize);
        handleResize();
        return () => window.removeEventListener('resize', handleResize);
    }, []);

    return windowSize;
}

const ItemComponent = ({component}) => <Async component={component} />;

const Portal: StyledType = ({ classes, children }) => {
    const {tabs = [], menu = [], tabIndex = 0} = useSelector(state => state.portal || {});
    const dispatch = useDispatch();
    const location = useLocation();
    const size = useWindowSize();
    const {portalName} = React.useContext(Context);
    const handleClick = React.useCallback(({itemData}) => itemData.page && dispatch({
        type: 'front.tab.show',
        component: itemData.page,
        title: itemData.title,
        path: '/' + itemData.page.name
    }), [dispatch]);
    const handleTabSelect = React.useCallback(({name, value}) => {
        (name === 'selectedIndex') && dispatch({
            type: 'front.tab.switch',
            tabIndex: value
        });
    }, [dispatch]);
    if (location.pathname !== '/' && !tabs.find(tab => tab.path === location.pathname)) {
        dispatch({
            type: 'front.route.find',
            path: location.pathname
        });
    }
    return (
        <Box direction='col' width='100%' height={size.height}>
            <Item baseSize={59}>
                <div className={classes.headerContainer}>
                    <Box direction='row' height='100%' crossAlign='center'>
                        <Item baseSize={46}>
                            <div className={classes.headerLogo}></div>
                        </Item>
                        <Item baseSize='auto'>
                            <div className={classes.headerTitle}>
                                <Text>{portalName}</Text>
                            </div>
                        </Item>
                        <Item baseSize='auto'>
                            <Menu dataSource={menu}
                                displayExpr='title'
                                showFirstSubmenuMode='onClick'
                                orientation='horizontal'
                                submenuDirection='auto'
                                hideSubmenuOnMouseLeave={false}
                                onItemClick={handleClick}
                            />
                        </Item>
                    </Box>
                </div>
            </Item>
            <Item ratio={1}>
                <TabPanel
                    height='100%'
                    dataSource={tabs}
                    selectedIndex={tabIndex}
                    itemTitleComponent={Title}
                    repaintChangesOnly
                    itemRender={ItemComponent}
                    onOptionChanged={handleTabSelect}
                />
                {children}
            </Item>
        </Box>
    );
};

export default Styled(Portal);
