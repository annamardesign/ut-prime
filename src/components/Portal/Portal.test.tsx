import React from 'react';
import {Provider} from 'react-redux';
import {createStore} from 'redux';

import { render } from '../test';
import Portal from './index';

describe('<Portal />', () => {
    it('render equals snapshot', () => {
        const store = createStore(() => ({
            login: {
                get(name) {
                    return {
                        result: true
                    }[name];
                }
            }
        }));
        const { getByTestId } = render(
            <Provider store={store}>
                <Portal />
            </Provider>,
            {}
        );
        expect(getByTestId('ut-front-test')).toMatchSnapshot();
    });
});
