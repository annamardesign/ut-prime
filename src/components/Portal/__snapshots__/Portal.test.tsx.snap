// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Portal /> render equals snapshot 1`] = `
<div
  data-testid="ut-front-test"
>
  <div
    class="p-d-flex p-flex-column"
    style="height: 768px;"
  >
    <div
      class="Portal-headerContainer-1"
    >
      <div
        class="p-d-flex p-ai-center p-jc-center"
      >
        <div
          class="Portal-headerLogo-3"
        />
        <div
          class="Portal-headerTitle-2"
        >
          <span>
            Administration Portal
          </span>
        </div>
        <div
          class="p-menubar p-component Portal-menuGrow-5"
        >
          <a
            class="p-menubar-button"
            href="#"
            role="button"
            tabindex="0"
          >
            <i
              class="pi pi-bars"
            />
          </a>
          <ul
            class="p-menubar-root-list"
            role="menubar"
          />
        </div>
        <div
          class="p-menubar p-component Portal-menu-6"
        >
          <a
            class="p-menubar-button"
            href="#"
            role="button"
            tabindex="0"
          >
            <i
              class="pi pi-bars"
            />
          </a>
          <ul
            class="p-menubar-root-list"
            role="menubar"
          >
            <li
              class="p-menuitem"
              role="none"
            >
              <a
                aria-haspopup="true"
                class="p-menuitem-link"
                href="#"
                role="menuitem"
              >
                <span
                  class="p-menuitem-icon user"
                />
                <span
                  class="p-menuitem-text"
                >
                  N/A
                </span>
                <span
                  class="p-submenu-icon pi pi-angle-down"
                />
              </a>
              <ul
                class="p-submenu-list"
                role="menu"
              >
                <li
                  class="p-menuitem"
                  role="none"
                >
                  <a
                    aria-haspopup="false"
                    class="p-menuitem-link"
                    href="#"
                    role="menuitem"
                  >
                    <span
                      class="p-menuitem-text"
                    >
                      Logout
                    </span>
                  </a>
                </li>
              </ul>
            </li>
          </ul>
        </div>
      </div>
    </div>
    <div
      class="p-tabview p-component Portal-tabs-4"
    >
      <ul
        class="p-tabview-nav"
        role="tablist"
      >
        <li
          class="p-tabview-ink-bar"
          style="width: 0px;"
        />
      </ul>
      <div
        class="p-tabview-panels"
      />
    </div>
  </div>
</div>
`;
